### PR TITLE
Improve file access operations and add NixOS path

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

File: lxd/instance_exec.go [hunk 1]
```diff
@@ -519,8 +519,15 @@ func instanceExecPost(d *Daemon, r *http.Request) response.Response {
 	_, ok := post.Environment["PATH"]
 	if !ok {
 		post.Environment["PATH"] = "/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
-		if inst.FileExists("/snap") == nil {
-			post.Environment["PATH"] = fmt.Sprintf("%s:/snap/bin", post.Environment["PATH"])
+
+		// Add some additional paths. This directly looks through /proc
+		// rather than use FileExists as none of those paths are expected to be
+		// symlinks and this is much faster than forking a sub-process and
```
Comment (mkg20001) [inline lxd/instance_exec.go:525]: /run/current-system is a symlink into /nix/store, not sure if that'll work

Comment (mkg20001) [reply lxd/instance_exec.go:525]: checking whether /run/current-system exists (but checking so it ignores the fact the symlink points to an invalid location) should fix that

could be key->value 

thingToCheck = path

/snap = /snap/bin
/run/current-system = /run/current-system/sw/bin

Comment (stgraber) [reply lxd/instance_exec.go:525]: We certainly can make a map of some kind for those checks, though if we can point directly to the absolute target of the symlink, that's likely a better option which will also save a tiny bit of effort to the shell later on.

Comment (mkg20001) [reply lxd/instance_exec.go:525]: Actually /run/current-system doesn't link to /nix/store but to /nix/store/unique-hash and that unique hash changes everytime, so the best we can do is to check /run/current-system, as everything else is variable

Another idea is to just check if /etc/NIXOS (with exactly that case) exists which is the marker nixOS uses internally to identify if this system is a NixOS installation

(Checking for /nix/store is also possible but nix can be installed independently of nixOS so we'd detect /nix/store on say ubuntu with nix installed aswell)

Comment (stgraber) [reply lxd/instance_exec.go:525]: Ok, changed to a map which maps the path to check to the path to add.

**Review by mkg20001 — APPROVED** [review]
LGTM

**Comment by stgraber** [comment]
@tomponline test results looking good

**Review by tomponline — APPROVED** [review]

File: lxd/instance_exec.go [hunk 2]
```diff
@@ -519,8 +519,21 @@ func instanceExecPost(d *Daemon, r *http.Request) response.Response {
...
+		// attaching to the instance.
+
+		extraPaths := map[string]string{
+			"/snap":      "/snap/bin",
+			"/etc/NIXOS": "/run/current-system/sw/bin",
+		}
+
+		for k, v := range extraPaths {
+			if shared.PathExists(fmt.Sprintf("/proc/%d/root%s", inst.InitPID(), k)) {
```
Comment (tomponline) [inline lxd/instance_exec.go:534]: @stgraber Nice approach! :)

**Pull request merged**